cc/crypto/hpke/sender_context.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2023 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cc/crypto/hpke/sender_context.h"
+
+#include <memory>
+#include <vector>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "openssl/hpke.h"
+
+namespace oak::crypto {
+
+namespace {
+constexpr size_t kAeadAlgorithmKeySizeBytes = 32;
+constexpr size_t kAeadNonceSizeBytes = 12;
+
+// Converts a string view object to a vector of bytes.
+std::vector<uint8_t> StringViewToBytes(absl::string_view str_view) {
+  std::vector<uint8_t> byte_array;
+  for (char character : str_view) {
+    byte_array.push_back(character);
+  }
+  return byte_array;
+}
+}  // namespace
+
+absl::StatusOr<std::string> SenderRequestContext::Seal(absl::string_view plaintext,
+                                                       absl::string_view associated_data) {
+  std::vector<uint8_t> plaintext_bytes = StringViewToBytes(plaintext);
+  std::vector<uint8_t> associated_data_bytes = StringViewToBytes(associated_data);
+  size_t max_out_len = EVP_HPKE_CTX_max_overhead(hpke_context_.get()) + plaintext_bytes.size();
+
+  std::vector<uint8_t> encrypted_message(max_out_len);
+  size_t encrypted_message_len;
+  if (!EVP_HPKE_CTX_seal(
+          /* ctx= */ hpke_context_.get(),
+          /* out= */ encrypted_message.data(),
+          /* out_len= */ &encrypted_message_len,
+          /* max_out_len= */ max_out_len,
+          /* in= */ plaintext_bytes.data(),
+          /* in_len= */ plaintext_bytes.size(),
+          /* ad= */ associated_data_bytes.data(),
+          /* ad_len= */ associated_data_bytes.size())) {
+    return absl::AbortedError("Failed to seal request");
+  }
+  encrypted_message.resize(encrypted_message_len);
+
+  std::string serialized_encrypted_message(encrypted_message.begin(), encrypted_message.end());
```

Comment:
I think if the previous variable will be renamed to `ciphertext_bytes`, this one can just be `ciphertext`, to avoid confusion with the serialized encrypted message proto

Reply:
sg